notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+    "# saving all results for comparison \n",
+    "join_results.append({\"query_type\":\"subquery_prejoin\",\"results\":result})\n",
+    "print_timestamp()\n",
+    "print(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "64372aea-9915-4e8a-a4ec-b34b160bd3b3",
+   "metadata": {},
+   "source": [
+    "### Results by Join Strategy"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "id": "81807f87-cff5-4e1c-8dd7-cc26b2b169e6",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[{'query_type': 'simple_join',\n",
+       "  'results': 'Results = avg:1182.9867 ms   min:1093.643 ms  max:1455.442 ms'},\n",
+       " {'query_type': 'simple_join_lookup',\n",
+       "  'results': 'Results = avg:254.4789 ms   min:234.103 ms  max:449.486 ms'},\n",
+       " {'query_type': 'simple_prejoin',\n",
+       "  'results': 'Results = avg:128.40235 ms   min:122.407 ms  max:176.861 ms'},\n",
+       " {'query_type': 'subquery_join',\n",
+       "  'results': 'Results = avg:2959.44585 ms   min:2841.148 ms  max:3093.195 ms'},\n",
+       " {'query_type': 'subquery_join_lookup',\n",
+       "  'results': 'Results = avg:550.8509 ms   min:519.703 ms  max:608.721 ms'},\n",
+       " {'query_type': 'subquery_prejoin',\n",
+       "  'results': 'Results = avg:352.1546 ms   min:327.998 ms  max:471.693 ms'}]"
+      ]
+     },
+     "execution_count": 22,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# review the results\n",
+    "sorted_list = sorted(join_results, key=lambda x: x['query_type'])\n",
+    "sorted_list"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "66fd0a23-cc14-4537-8965-5dab8eeb1b46",
+   "metadata": {},
+   "source": [
+    "In these results you can see the progression of response times. \n",
+    "- The use of lookups improves performance over joining to another datasource, but it requires enough Heap to fit all lookups on all historical and ingestion tasks (Peon) JVMs.\n",
+    "- The pre-joined table performs even better and has the added advantage of not requiring any additional Heap.\n",
+    "\n",
+    "If the data being joined changes over time and you wish to query with the latest values, consider a lookup. If the data does not change or the change is not desirable, then pre-join the data. The example of `age_group` as the joined property works well in the pre-join strategy because the user's age_group is important at the time they clicked, not when looking at old click rows based on their current age_group."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "46e2e6f6-86e1-48dd-a125-6c1920173d74",
+   "metadata": {},
+   "source": [
+    "## Asynchronous Queries on MSQ Engine\n",
+    "\n",
+    "The MSQ (Multi-stage Query Engine) engine works by shuffling rows between stages. All stages execute in tasks in parallel as much as they can.\n",
+    "![](assets/msq-engine.png)\n",
```

Comment:
Documentation.

Reply:
Where? I don't see explanations about the MSQ engine processing model or how broadcast or sortMerge work, just how to specify them.... I think this is useful for people to understand.